python/cudf/cudf/core/join/casting_logic.py

```diff
@@ -0,0 +1,199 @@
+    proceed. This function handles categorical variables.
+    Categorical variable typecasting logic depends on both `how`
+    and the specifics of the categorical variables to be merged.
+    Merging categorical variables when only one side is ordered
+    is ambiguous and not allowed. Merging when both categoricals
+    are ordered is allowed, but only when the categories are
+    exactly equal and have equal ordering, and will result in the
+    common dtype.
+    When both sides are unordered, the result categorical depends
+    on the kind of join:
+    - For inner joins, the result will be the intersection of the
+    categories
+    - For left or right joins, the result will be the the left or
+    right dtype respectively. This extends to semi and anti joins.
+    - For outer joins, the result will be the union of categories
+    from both sides.
+
+    """
+    ltype = lcol.dtype
+    rtype = rcol.dtype
+
+    # this function is only to be used to resolve the result when both
+    # sides are categorical
+    if not isinstance(ltype, CategoricalDtype) and isinstance(
+        rtype, CategoricalDtype
+    ):
+        raise TypeError("Both operands must be CategoricalDtype")
+
+    # true for every configuration
+    if ltype == rtype:
+        return ltype
+
+    # raise for any join where ordering doesn't match
+    if ltype.ordered != rtype.ordered:
+        raise TypeError(
+            "Merging on categorical variables with mismatched"
+            " ordering is ambiguous"
+        )
+    elif ltype.ordered and rtype.ordered:
+        # if we get to here, categories must be what causes the
+        # dtype equality check to fail. And we can never merge
+        # two ordered categoricals with different categories
+        raise TypeError(
+            f"{how} merge between categoricals with "
+            "different categories is only valid when "
+            "neither side is ordered"
+        )
+
+    elif how == "inner":
+        # neiter ordered, so categories must be different
+        # demote to underlying types
+        return _input_to_libcudf_castrules_any(
+            ltype.categories, rtype.categories, how
+        )
+
+    elif how == "left":
+        return ltype
+    elif how == "right":
+        return rtype
+
+    elif how == "outer":
+        new_cats = cudf.concat([ltype.categories, rtype.categories]).unique()
+        return cudf.CategoricalDtype(categories=new_cats, ordered=False)
+
+
+def _input_to_libcudf_castrules_any_cat(lcol, rcol, how):
+
+    l_is_cat = isinstance(lcol.dtype, CategoricalDtype)
+    r_is_cat = isinstance(rcol.dtype, CategoricalDtype)
+
+    if l_is_cat and r_is_cat:
+        return _input_to_libcudf_castrules_both_cat(lcol, rcol, how)
+    elif l_is_cat or r_is_cat:
+        return (
+            lcol.dtype.categories.dtype
+            if l_is_cat
+            else rcol.dtype.categories.dtype
+        )
+    else:
+        raise ValueError("Neither operand is categorical")
```

Comment:
having a docstring here and in the public method to capture this behavior as well would be good. If I understand correctly the behavior is if merging a categorical against a non-categorical, the return type is always non-categorical?

Reply:
Yeah it is. Although now that I think about it, maybe it should be categorical at least for left joins if the left dtype is categorical.

Reply:
So the behavior would be:
- left:
  - left is categorical: categorical encode based on left
  - right is categorical: categorical encode based on right?
- right:
  - left is categorical: categorical encode based on left?
  - right is categorical: categorical encode based on right
- inner: decategorize?
- outer: decategorize

I'm unclear on the cases with question marks.

Reply:
I'd suggest to raise in the `?` cases for left and right. The reason is that a left join seems like it's meant to take an existing dataset and add information to it from another dataset in the form of new columns, or possibly extra rows if there's duplicated matches. Maybe it's just me but it seems wrong to propagate the categorical from the right hand dataset into the result. 

For inner and outer joins where both tables are on the same footing I think it's best not to create a categorical variable for the user that would be the subset or superset of categories. Whereas if both sides are categorical from the start, it's a fair assumption that the user wants a categorical at the end.

Reply:
The other thing to note is that we currently prioritize the left dtype if possible in left joins and right dtype in right joins, originally for the same reason as above.

Reply:
Okay, so if I understand properly this is the proposal:
- left:
  - left is categorical, right is not: categorical encode based on left
  - right is categorical, left is not: throw an exception
- right:
  - left is categorical, right is not: throw an exception
  - right is categorical, left is not: categorical encode based on right
- inner:
  - left is categorical, right is not: decategorize
  - right is categorical, left is not: decategorize
- outer:
  - left is categorical, right is not: decategorize
  - right is categorical, left is not: decategorize

Reply:
Yes, that's the proposal, although I'm slightly more in support of the approach where we do not decategorize in the left and right join non-exception cases.

Reply:
Implemented as described above.